src/internet_identity/src/storage/tests.rs

```diff
@@ -0,0 +1,175 @@
+use crate::storage::{Header, StorageError};
+use crate::{DeviceDataInternal, Storage};
+use ic_stable_structures::{Memory, VectorMemory};
+use internet_identity_interface::{DeviceProtection, KeyType, Purpose};
+use serde_bytes::ByteBuf;
+
+const HEADER_SIZE: usize = 58;
+
+#[test]
+fn should_match_actual_header_size() {
+    // if this test fails, make sure the change was intentional and upgrade as well as rollback still work!
+    assert_eq!(std::mem::size_of::<Header>(), HEADER_SIZE);
+}
+
+#[test]
+fn should_report_max_number_of_entries_for_8gb() {
+    let memory = VectorMemory::default();
+    let storage = Storage::<(), VectorMemory>::new((1, 2), memory);
+    assert_eq!(storage.max_entries(), 3774873);
+}
+
+#[test]
+fn should_serialize_header() {
+    let memory = VectorMemory::default();
+    let mut storage = Storage::<(), VectorMemory>::new((1, 2), memory.clone());
+    storage.update_salt([5u8; 32]);
+    storage.flush();
+
+    let mut buf = vec![0; HEADER_SIZE];
+    memory.read(0, &mut buf);
+    assert_eq!(buf, hex::decode("49494301000000000100000000000000020000000000000000080505050505050505050505050505050505050505050505050505050505050505").unwrap());
+}
+
+#[test]
+fn should_recover_header_from_memory() {
+    let memory = VectorMemory::default();
+    memory.grow(1);
+    memory.write(0, &hex::decode("494943010500000040e2010000000000f1fb09000000000000084343434343434343434343434343434343434343434343434343434343434343").unwrap());
```

Comment:
I'm guessing that somewhere in there there's the storage range 123456 - 654321. Did you run the test and then copy out the hex data? Lots of these tests seem to be testing a particular encoding of the header, and sometimes (like here) a particular field value. What do these tests try to prevent in terms of bugs?

Reply:
Yes. These tests detect if the memory location shifts or the data layout changes, which are both things that must not happen by accident.